Checkbox: implement ControlValueAccessor to trigger changes within Angular forms

AgreSanGit opened the pull request:
## Which issue does this PR close?

This PR closes #2804 

## What is the new behavior?

ControlValueAccessor implemented so kirby checkbox does now trigger changes within reactive forms

<!-- Replace this paragraph with a description of the new behaviour after your pull request is merged -->

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, replace this paragraph with a description of the impact and migration path for existing applications  -->

## Are there any additional context?

<!-- Replace this paragraph with any additional context e.g, explanations, links or screenshots (if any) -->

## Checklist:

The following tasks should be carried out in sequence in order to follow [the process of contributing](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#the-process-of-contributing) correctly.

### Reminders
- [ ] Make sure you have implemented tests following the guidelines in: "[The good: Test](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Test)".
- [ ] Make sure you have updated the cookbook with examples and showcases (for bug fixes, enhancements & new components).

### Review  
- [x] Determine if your changes are a fix, feature or breaking-change, and add the matching label to your PR. If it is tooling, dependency updates or similar, add ignore-for-release.
- [x] Do a [self-review](https://github.com/kirbydesign/designsystem/wiki/The-Good%3A-Self-review).
- [x] Request that the changes are code-reviewed 
- [x] Request that the changes are [UX reviewed](https://github.com/kirbydesign/designsystem/blob/main/.github/CONTRIBUTING.md/#ux-review) (only necessary if your PR introduces visual changes)

When the pull request has been approved it will be merged to `develop` by Team Kirby.

AgreSanGit committed:
Implement ControlValueAcessor

AgreSanGit committed:
ControlValueAccessor test

AgreSanGit committed:
Lint error because of private method

AgreSanGit committed:
Changes to test to remove ts ignore

mark-drastrup commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,47 @@ export class CheckboxComponent {
+
+  //ControleValueAcessor
+  private onChange: (value: boolean) => void = () => {};
+  private onTouched: () => void = () => {};
```
```suggestion
  private _onTouched = () => {};
```

mark-drastrup replied:
This is just to follow the naming convention in other components, that also implements the ControlValueAccessor interface :)

mark-drastrup commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,47 @@ export class CheckboxComponent {
   _labelId = UniqueIdGenerator.scopedTo('kirby-checkbox-label').next();
+
+  //ControleValueAcessor
+  private onChange: (value: boolean) => void = () => {};
```
```suggestion
  private _onChange: (value: boolean) => void = () => {};
```

mark-drastrup replied:
This is just to follow the naming convention in other components, that also implements the ControlValueAccessor interface :)

mark-drastrup commented on libs/designsystem/checkbox/src/checkbox.component.spec.ts:
```diff
@@ -202,4 +202,32 @@ describe('CheckboxComponent', () => {
       expect(spectator.component.checkedChange.emit).toHaveBeenCalledWith(false);
     });
   });
+  describe('ControlValueAccessor implementation', () => {
```
```suggestion
  describe('implementing ControlValueAccessor interface', () => {
```

mark-drastrup replied:
This is not important, but it's just the description that is being used in other test files, that tests for ControlCalueAccessor implementation :)

mark-drastrup commented on libs/designsystem/checkbox/src/checkbox.component.spec.ts:
```diff
@@ -202,4 +202,32 @@ describe('CheckboxComponent', () => {
+      const newValue = true;
+      spectator.component.registerOnChange(onChangeSpy);
+      spectator.component.onChecked(newValue);
+      expect(onChangeSpy).toHaveBeenCalledWith(newValue);
```
```suggestion
      expect(onChangeSpy).toHaveBeenCalledOnceWith(newValue);
```

mark-drastrup replied:
This is just to make sure that the onChange function is only called once :)

mark-drastrup commented on libs/designsystem/checkbox/src/checkbox.component.spec.ts:
```diff
@@ -202,4 +202,32 @@ describe('CheckboxComponent', () => {
+      const onTouchedSpy = jasmine.createSpy('onTouched');
+      spectator.component.registerOnTouched(onTouchedSpy);
+      spectator.component.onChecked(true);
+      expect(onTouchedSpy).toHaveBeenCalled();
```
```suggestion
      expect(onTouchedSpy).toHaveBeenCalled(); 
      expect(onTouchedSpy).toHaveBeenCalledTimes(1);
```

mark-drastrup replied:
Again just to make sure that the onTouched function is only called once :)

mark-drastrup reviewed (dismissed):
Great work @AgreSanGit :D I have have some nitty gritty comments! :)

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: mark-drastrup <mark.drastrup@gmail.com>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: mark-drastrup <mark.drastrup@gmail.com>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.spec.ts

Co-authored-by: mark-drastrup <mark.drastrup@gmail.com>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.spec.ts

Co-authored-by: mark-drastrup <mark.drastrup@gmail.com>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.spec.ts

Co-authored-by: mark-drastrup <mark.drastrup@gmail.com>

AgreSanGit committed:
Pretter/Lint error correction after code review

AgreSanGit committed:
Prettier/Lint correction

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

RasmusKjeldgaard commented on libs/designsystem/checkbox/src/checkbox.component.spec.ts:
```diff
@@ -202,4 +202,33 @@ describe('CheckboxComponent', () => {
+      const onTouchedSpy = jasmine.createSpy('onTouched');
+      spectator.component.registerOnTouched(onTouchedSpy);
+      spectator.component.onChecked(true);
+      expect(onTouchedSpy).toHaveBeenCalled();
```
```suggestion
```

I suggest to remove this line to avoid redundant expects, as it is covered by `toHaveBeenCalledTimes(1)` on the next line

RasmusKjeldgaard commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,49 @@ export class CheckboxComponent {
   // IDs used for a11y labelling
   _labelId = UniqueIdGenerator.scopedTo('kirby-checkbox-label').next();
+
+  //ControleValueAcessor
```
```suggestion
  // Initialize empty ControlValueAccessor callback functions
```
Consider giving a bit of context in this comment

RasmusKjeldgaard commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,49 @@ export class CheckboxComponent {
+   */
+  setDisabledState(isDisabled: boolean): void {
+    this.disabled = isDisabled;
+  }
```
Consider just copying this whole block directly from `dropdown.component.ts` (except for the implementation of writeValue of course) for consistency of comments and so on.

RasmusKjeldgaard reviewed (changes requested):
Good job on implementation and test!

Apart from a couple of minor suggestions, I think the only thing missing is a good example of this in a reactive form, like we have for radio-groups: https://cookbook.kirby.design/#/home/showcase/radio

RasmusKjeldgaard commented:
I think we should label this as a feature instead of a fix. I know the original issue is a bug report, but we are adding a new feature cleanly here, not fixing existing functionality.

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.spec.ts

Co-authored-by: RasmusKjeldgaard <rkk@bankdata.dk>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: RasmusKjeldgaard <rkk@bankdata.dk>

AgreSanGit committed:
Adjust controlValueAccesor to be consistent

AgreSanGit committed:
Add example and showcase for ControleValueAccessor

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Update codeSnippet

AgreSanGit committed:
Merge branch 'bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms' of https://github.com/kirbydesign/designsystem into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Remove template driven from showcase

RasmusKjeldgaard commented on apps/cookbook/src/app/showcase/checkbox-showcase/checkbox-showcase.component.html:
```diff
@@ -70,6 +70,27 @@ <h2>Checked Change Event</h2>
     <cookbook-checkbox-events-example #checkboxEventsExample></cookbook-checkbox-events-example>
   </cookbook-example-viewer>
+  <h2>ngForms</h2>
+  <h3>Reactive Forms</h3>
```
```suggestion
  <h2>Angular Forms</h2>
```

RasmusKjeldgaard replied:
I still think we can make this a bit easier to digest, possibly by adding back a little bit of the text inspired by the radio group example right below the Angular Forms heading:

> The checkbox can be used directly in [template driven](https://angular.io/guide/forms) and [reactive forms](https://angular.io/guide/reactive-forms).
> To use checkbox in a reactive form, bind your form control to <kirby-checkbox> with simple data binding:

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/checkbox-example/examples/reactive-forms.ts:
```diff
@@ -0,0 +1,79 @@
+  formControlName="isChecked" 
+  text="Reactive Forms" 
+  [disabled]="!canSelectFavorite">
+  </kirby-checkbox>
```
Consider adding a couple of checkboxes more, and reuse text from Radio Examples: 

<img width="180" alt="image" src="https://github.com/kirbydesign/designsystem/assets/42470636/312fd969-79d2-442f-82f8-67d233e4ef4a">

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/checkbox-example/checkbox-example.module.ts:
```diff
@@ -10,7 +11,7 @@ import { CheckboxListExampleComponent } from './examples/list';
 import { CheckboxSizesExampleComponent } from './examples/sizes';
 import { CheckboxStatesExampleComponent } from './examples/states';
-
+import { CheckboxReativeFormsExampleComponent } from './examples/reactive-forms';
```
```suggestion
import { CheckboxReactiveFormsExampleComponent } from './examples/reactive-forms';

```

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/checkbox-example/examples/reactive-forms.ts:
```diff
@@ -0,0 +1,79 @@
+  [disabled]="!canSelectFavorite">
+  </kirby-checkbox>
+</form>
+<fieldset>
```
Consider using `cookbook-example-configuration-wrapper` here.

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/checkbox-example/examples/reactive-forms.ts:
```diff
@@ -0,0 +1,79 @@
+      this.form.get('isChecked')?.disable();
+    }
+  }
+    `,
```
```suggestion
toggleCanSelectFavorite() {
  this.canSelectFavorite = !this.canSelectFavorite;

  if (this.canSelectFavorite) {
    this.form.get('isChecked')?.enable();
  } else {
    this.form.get('isChecked')?.disable();
  }
}
`,
```

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/checkbox-example/examples/reactive-forms.ts:
```diff
@@ -0,0 +1,79 @@
+  codeSnippet: string = config.codeSnippet;
+  form: FormGroup;
+  canSelectFavorite = true;
+  favoriteRequired = true;
```
Consider removing `favoriteRequired` if it is not used in the example

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -3,12 +3,14 @@ import {
   Output,
 } from '@angular/core';
 import { IonicModule } from '@ionic/angular';
 import { UniqueIdGenerator } from '@kirbydesign/designsystem/helpers';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
```
Please consider readability: order and grouping of imports: 
```suggestion
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

import { UniqueIdGenerator } from '@kirbydesign/designsystem/helpers';
```

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,59 @@ export class CheckboxComponent {
   // IDs used for a11y labelling
   _labelId = UniqueIdGenerator.scopedTo('kirby-checkbox-label').next();
+
+  // Initialize empty ControlValueAccessor callback functions
```
Consider wording:
```suggestion
  // Initialize default ControlValueAccessor callback functions (noop)
```

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,59 @@ export class CheckboxComponent {
+  private _onTouched = () => {};
+
+  /**
+   * Sets the select's value. Part of the ControlValueAccessor interface
```
Copy pasted from somehere? ;)
```suggestion
   * Sets the checkbox's value. Part of the ControlValueAccessor interface
```

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,59 @@ export class CheckboxComponent {
+  }
+
+  /**
+   * Saves a callback function to be invoked when the select's value
```
Copy pasted from somehere? ;)
```suggestion
   * Saves a callback function to be invoked when the checkbox's value
```

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,59 @@ export class CheckboxComponent {
+  }
+
+  /**
+   * Saves a callback function to be invoked when the select is blurred
```
Copy pasted from somehere? ;)
```suggestion
   * Saves a callback function to be invoked when the checkbox is blurred
```

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,59 @@ export class CheckboxComponent {
+  }
+
+  /**
+   * Disables the select. Part of the ControlValueAccessor interface required
```
Copy pasted from somehere? ;)
```suggestion
   * Disables the checkbox. Part of the ControlValueAccessor interface required
```

jakobe commented on libs/designsystem/checkbox/src/checkbox.component.ts:
```diff
@@ -51,9 +60,59 @@ export class CheckboxComponent {
     this.checked = checked;
+    this._onChange(this.checked);
     this.checkedChange.emit(this.checked);
+    this._onTouched();
```
This should be called on blur, not when checked state changes.

jakobe reviewed (changes requested):
Hi @AgreSanGit ,

1 important comment and a few copy-pastes to fix ;)

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: Jakob Engelbrecht <jakob@basher.dk>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: Jakob Engelbrecht <jakob@basher.dk>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: Jakob Engelbrecht <jakob@basher.dk>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: Jakob Engelbrecht <jakob@basher.dk>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: Jakob Engelbrecht <jakob@basher.dk>

AgreSanGit committed:
Update apps/cookbook/src/app/showcase/checkbox-showcase/checkbox-showcase.component.html

Co-authored-by: RasmusKjeldgaard <rkk@bankdata.dk>

AgreSanGit committed:
Update apps/cookbook/src/app/examples/checkbox-example/checkbox-example.module.ts

Co-authored-by: RasmusKjeldgaard <rkk@bankdata.dk>

AgreSanGit committed:
Update apps/cookbook/src/app/examples/checkbox-example/examples/reactive-forms.ts

Co-authored-by: RasmusKjeldgaard <rkk@bankdata.dk>

AgreSanGit committed:
Update libs/designsystem/checkbox/src/checkbox.component.ts

Co-authored-by: Jakob Engelbrecht <jakob@basher.dk>

AgreSanGit committed:
Fix spelling error

AgreSanGit committed:
onTouched on blur instead of checked state changes

AgreSanGit committed:
Update cookbook example

AgreSanGit committed:
CSS display block fix

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
move ng-forms.shared.scss and share it

AgreSanGit committed:
Links to template driven and reactive forms doc

RasmusKjeldgaard committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

RasmusKjeldgaard commented on apps/cookbook/src/app/examples/checkbox-example/examples/reactive-forms.ts:
```diff
@@ -0,0 +1,135 @@
+    </p>
+  </section>
+  </cookbook-example-configuration-wrapper>`,
+  codeSnippet: `checkboxes: { id: string; label: string }[] = [
+    { id: 'Bacon', label: 'Bacon' },
+    { id: 'Salami', label: 'Salami' },
+    { id: 'Tenderloin', label: 'Tenderloin' },
+  ];
+  
+  form: FormGroup;
+  canSelectFavorite = true;
+  formValue: any;
```
I think we should clean this example up, so it is easier to understand.
Let us talk about this thursday!

AgreSanGit committed:
Checkbox reactive exmaple simplfied significantly

AgreSanGit committed:
Merge branch 'bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms' of https://github.com/kirbydesign/designsystem into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Add Configuration wrapper for forms

AgreSanGit committed:
Adjust example configuration wrapper reactive forms

AgreSanGit committed:
Merge remote-tracking branch 'origin/develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Fix double providers after merge

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Rename cookbook reactive form state

AgreSanGit committed:
Empty div

AgreSanGit committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms

AgreSanGit committed:
Fix build issue

RasmusKjeldgaard committed:
Fix whitespace issues

RasmusKjeldgaard committed:
Remove unique labelID

This is no longer needed and was removed in another PR but seems to have
been accidentally added back

RasmusKjeldgaard committed:
Merge branch 'develop' into bug/2804-kirby-checkbox-does-not-trigger-changes-within-reactive-forms
